File: Monitor_c/ZCR.c
#include "zcro.h"
#include "mean.h"
#include "enframe.h"

int ZCR(double data[], int size, int fs, int nBit)
{
	int i;
	double num = 0, sum =0;
	double secPerFra = 20*0.001;	
	int frameLength = secPerFra*fs;	//frame length
	int frameInc = frameLength/2;	//frame increment
	double **frames;
	int frameNum;
	double *zcr;
	double feature_ZCR;

	enframe(data,size,frameLength,frameInc,&frames,&frameNum);

	zcro(frames,frameNum,frameLength,&zcr);
	

	for(i=0;i<frameNum;i++)
		if(zcr[i]>0)
		{
			sum += zcr[i];
			num = num + 1 ;
		}

	feature_ZCR = sum/num;

	return feature_ZCR;
}

File: matlab/ZCR.m
function [ feature_ZCR ] = ZCR( voice, fs, nBit )
%ZCR Calc Zero-crossing Rate
%   input: voice data
%   input: sample rate
%   input: number of bits per sample
%   output: zero-cross rate feature of the voice data

    disp('Feature ZCR');
    frameLength = 20*0.001*fs; %frame length
    frameInc = frameLength/2;   %frame inc
    voiceFrames = enframe(voice, frameLength, frameInc);
    
    zcr=zcro(voiceFrames);
    feature_ZCR = mean(zcr(zcr~=0));
    
%     plot(zcr);
%     disp(mean(zcr));
%     xlabel('Frame Num');
%     ylabel('ZCR');
%     title(mean(zcr));

end

